File: asr/server/src/metrics.cc
#include "metrics.h"

#include <iostream>

namespace asr_service {

namespace {

const prometheus::Summary::Quantiles kQuantiles = {
    {0.5,  0.05},
    {0.9,  0.01},
    {0.99, 0.001},
};

}  // namespace

Metrics::Metrics(const std::string &listen, const std::string &model_id)
    : registry_(std::make_shared<prometheus::Registry>()), model_id_(model_id) {
    fam_active_ = &prometheus::BuildGauge()
                       .Name("asr_active_sessions")
                       .Help("Number of active streaming sessions")
                       .Register(*registry_);
    fam_partial_ = &prometheus::BuildCounter()
                        .Name("asr_partial_total")
                        .Help("Number of partial results emitted")
                        .Register(*registry_);
    fam_final_ = &prometheus::BuildCounter()
                      .Name("asr_final_total")
                      .Help("Number of final results emitted")
                      .Register(*registry_);
    fam_rtf_ = &prometheus::BuildSummary()
                    .Name("asr_rtf")
                    .Help("Real-time factor (decode_time / audio_duration)")
                    .Register(*registry_);
    fam_decode_ = &prometheus::BuildSummary()
                       .Name("asr_decode_latency_ms")
                       .Help("Per-chunk decode latency in milliseconds")
                       .Register(*registry_);
    fam_error_ = &prometheus::BuildCounter()
                      .Name("asr_error_total")
                      .Help("Number of errors broken down by code")
                      .Register(*registry_);

    active_sessions_   = &fam_active_->Add({{"model_id", model_id}});
    partial_total_     = &fam_partial_->Add({{"model_id", model_id}});
    final_total_       = &fam_final_->Add({{"model_id", model_id}});
    rtf_               = &fam_rtf_->Add({{"model_id", model_id}}, kQuantiles);
    decode_latency_ms_ = &fam_decode_->Add({{"model_id", model_id}}, kQuantiles);

    if (!listen.empty()) {
        try {
            exposer_ = std::make_unique<prometheus::Exposer>(listen);
            exposer_->RegisterCollectable(registry_);
            std::cerr << "[metrics] exposing on " << listen << "/metrics" << std::endl;
        } catch (const std::exception &e) {
            std::cerr << "[metrics] exposer failed: " << e.what() << std::endl;
        }
    }
}

prometheus::Counter &Metrics::error_total(int code) {
    return fam_error_->Add({
        {"model_id", model_id_},
        {"code", std::to_string(code)},
    });
}

}  // namespace asr_service

File: asr/ws-server/amphion_asr_ws/metrics.py
from __future__ import annotations

from prometheus_client import Counter, Gauge, Histogram, start_http_server

ACTIVE_SESSIONS = Gauge("asr_active_sessions", "Current active ASR WebSocket sessions")
PARTIAL_TOTAL = Counter("asr_partial_total", "Total partial ASR events", ["model_id"])
FINAL_TOTAL = Counter("asr_final_total", "Total final ASR events", ["model_id"])
ERROR_TOTAL = Counter("asr_error_total", "Total ASR errors", ["model_id", "code"])
DECODE_LATENCY_MS = Histogram(
    "asr_decode_latency_ms",
    "Batch decode latency in milliseconds",
    buckets=(1, 2, 5, 10, 20, 50, 100, 200, 500, 1000),
)
BATCH_SIZE = Histogram(
    "asr_decode_batch_size",
    "Observed decode batch size",
    buckets=(1, 2, 4, 8, 16, 32, 50, 64, 128),
)
QUEUE_DEPTH = Gauge("asr_decode_queue_depth", "Current decode queue depth")


def start_metrics_server(listen: str | None) -> None:
    if not listen:
        return
    host, port_text = listen.rsplit(":", 1)
    start_http_server(int(port_text), addr=host)
